src/librustdoc/clean/inline.rs

```diff
@@ -530,20 +529,18 @@ fn build_macro(cx: &mut DocContext<'_>, did: DefId, name: Symbol) -> clean::Item
     let imported_from = cx.tcx.crate_name(did.krate);
     match cx.enter_resolver(|r| r.cstore().load_macro_untracked(did, cx.sess())) {
         LoadedMacro::MacroDef(def, _) => {
-            let matchers: Vec<Span> = if let ast::ItemKind::MacroDef(ref def) = def.kind {
+            let matchers: Vec<_> = if let ast::ItemKind::MacroDef(ref def) = def.kind {
                 let tts: Vec<_> = def.body.inner_tokens().into_trees().collect();
-                tts.chunks(4).map(|arm| arm[0].span()).collect()
+                // FIXME: remove this clone
```

Comment:
I don't think this can be removed, I would just remove the comment.

Reply:
I found a way to get rid of the `clone` in `clean/mod.rs` and avoid collecting into a `Vec`, but I'm not sure if it's worth it. What do you think?

```diff
diff --git a/src/librustdoc/clean/inline.rs b/src/librustdoc/clean/inline.rs
index 613ac60a9ba..03304b4b981 100644
--- a/src/librustdoc/clean/inline.rs
+++ b/src/librustdoc/clean/inline.rs
@@ -531,7 +531,6 @@ fn build_macro(cx: &mut DocContext<'_>, did: DefId, name: Symbol) -> clean::Item
         LoadedMacro::MacroDef(def, _) => {
             let matchers: Vec<_> = if let ast::ItemKind::MacroDef(ref def) = def.kind {
                 let tts: Vec<_> = def.body.inner_tokens().into_trees().collect();
-                // FIXME: remove this clone
                 tts.chunks(4).map(|arm| arm[0].clone()).collect()
             } else {
                 unreachable!()
@@ -540,7 +539,7 @@ fn build_macro(cx: &mut DocContext<'_>, did: DefId, name: Symbol) -> clean::Item
             let source = format!(
                 "macro_rules! {} {{\n{}}}",
                 name.clean(cx),
-                utils::render_macro_arms(&matchers, ";")
+                utils::render_macro_arms(matchers.iter(), ";")
             );
 
             clean::MacroItem(clean::Macro { source, imported_from: Some(imported_from) })
diff --git a/src/librustdoc/clean/mod.rs b/src/librustdoc/clean/mod.rs
index 2949f502928..0b0ff34b71a 100644
--- a/src/librustdoc/clean/mod.rs
+++ b/src/librustdoc/clean/mod.rs
@@ -2161,11 +2161,10 @@ fn clean(&self, cx: &mut DocContext<'_>) -> Item {
         let name = renamed.unwrap_or(item.ident.name);
         let tts = item.ast.body.inner_tokens().trees().collect::<Vec<_>>();
         // Extract the macro's matchers. They represent the "interface" of the macro.
-        // FIXME: remove this clone
-        let matchers = tts.chunks(4).map(|arm| arm[0].clone()).collect::<Vec<_>>();
+        let matchers = tts.chunks(4).map(|arm| &arm[0]);
 
         let source = if item.ast.macro_rules {
-            format!("macro_rules! {} {{\n{}}}", name, render_macro_arms(&matchers, ";"))
+            format!("macro_rules! {} {{\n{}}}", name, render_macro_arms(matchers, ";"))
         } else {
             let vis = item.vis.clean(cx);
             let def_id = item.def_id.to_def_id();
@@ -2175,14 +2174,14 @@ fn clean(&self, cx: &mut DocContext<'_>) -> Item {
                     "{}macro {}{} {{\n    ...\n}}",
                     vis.to_src_with_space(cx.tcx, def_id),
                     name,
-                    matchers.iter().map(render_macro_matcher).collect::<String>(),
+                    matchers.map(render_macro_matcher).collect::<String>(),
                 )
             } else {
                 format!(
                     "{}macro {} {{\n{}}}",
                     vis.to_src_with_space(cx.tcx, def_id),
                     name,
-                    render_macro_arms(&matchers, ","),
+                    render_macro_arms(matchers, ","),
                 )
             }
         };
diff --git a/src/librustdoc/clean/utils.rs b/src/librustdoc/clean/utils.rs
index d90ee21cbd3..45973a7c194 100644
--- a/src/librustdoc/clean/utils.rs
+++ b/src/librustdoc/clean/utils.rs
@@ -551,7 +551,10 @@ fn print_const_with_custom_print_scalar(tcx: TyCtxt<'_>, ct: &'tcx ty::Const<'tc
 
 /// Render a sequence of macro arms in a format suitable for displaying to the user
 /// as part of an item declaration.
-pub(super) fn render_macro_arms(matchers: &[TokenTree], arm_delim: &str) -> String {
+pub(super) fn render_macro_arms<'a>(
+    matchers: impl Iterator<Item = &'a TokenTree>,
+    arm_delim: &str,
+) -> String {
     let mut out = String::new();
     for matcher in matchers {
         writeln!(out, "    {} => {{ ... }}{}", render_macro_matcher(matcher), arm_delim).unwrap();
```

Reply:
Sure, seems fine. You forgot to remove the first `.clone()` though :P

Reply:
Done! :)